### ISSUE TITLE
it is found that when the number of user groups is greater than 2, the content of the configmap user-groups-map or user-gids-map does not match the actual one.

### ISSUE THREAD

**Issue by 18707602767** [OSC/k8-ldap-configmap#20]
**Environment:**
- Linux: CentOS Linux release 7.8.2003 (Core)
- kubernetes: V1.24.2
- CRI: Containerd
- LDAP: FreeIPA

**What happened:**
The k8-ldap-configmap char is installed in the k8s cluster, and it is found that when the number of user's groups is greater than 2, the content of the configmap user-groups-map or user-gids-map does not match the actual one.

**values.yaml file**
```
# Default values for k8-ldap-configmap.
# This is a YAML-formatted file.
# Declare variables to be passed into your templates.

ldapUrl: 'ldap://10.49.118.102:389'
ldapTLS: false
ldapTLSVerify: true
ldapTLSCACert: ''
ldapGroupBaseDN: 'cn=groups,cn=accounts,dc=pl,dc=hpc'
ldapUserBaseDN: 'cn=users,cn=accounts,dc=pl,dc=hpc'
ldapBindDN: 'uid=rancher,cn=sysaccounts,cn=etc,dc=pl,dc=hpc'
ldapBindPassword: '1qavfsdfa'
ldapGroupFilter: '(objectClass=posixGroup)'
ldapUserFilter: '(objectClass=posixAccount)'
ldapMemberScheme: 'member'
ldapUserAttrMap: 'name=uid,uid=uidNumber,gid=gidNumber'
ldapGroupAttrMap: 'name=cn,gid=gidNumber'
mappers:
  - user-uid
  - user-gid
  - user-gids
  - user-groups
mappersUserFilter: []
mappersGroupFilter: []
userPrefix: ''
interval: 5s
# Set namespace of generated ConfigMaps
# Defaults to namespace of Chart release
namespaceConfigMap: "default"

extraArgs: []

image:
  repository: docker.io/ohiosupercomputer/k8-ldap-configmap
  pullPolicy: IfNotPresent
  # Overrides the image tag whose default is the chart appVersion.
  tag: ""

imagePullSecrets: []
nameOverride:
fullnameOverride:
namespace:

rbac:
  create: true
  serviceAccount:
    create: true
    # Annotations to add to the service account
    annotations: {}
    # The name of the service account to use.
    # If not set and create is true, a name is generated using the fullname template
    name: ""

podAnnotations: {}

podSecurityContext: {}
  # fsGroup: 2000

securityContext:
  allowPrivilegeEscalation: false
  capabilities:
    drop:
    - ALL
  seccompProfile:
    type: RuntimeDefault
  privileged: false
  readOnlyRootFilesystem: true
  runAsNonRoot: true
  runAsUser: 65534
  runAsGroup: 65534

service:
  # Specifies whether a service should be created
  create: true
  type: ClusterIP
  port: 8080
  annotations:
    prometheus.io/scrape: 'true'
    prometheus.io/path: /metrics

resources:
  limits:
    memory: 128Mi
  requests:
    cpu: 100m
    memory: 50Mi

nodeSelector: {}

tolerations: []

affinity: {}

```
**kubectl get configmap user-groups-map -o yaml**
```
apiVersion: v1
data:
  admin: '["admins"]'
  zhanrd: '["test-1","test-3"]'
kind: ConfigMap
metadata:
  creationTimestamp: "2023-01-09T09:08:41Z"
  name: user-groups-map
  namespace: default
  resourceVersion: "9170354"
  uid: 6fdb1e61-bd34-4cae-a886-a31864686d18

```
**id zhanrd**
```
uid=20000(zhanrd) gid=10000(test-1) groups=10000(test-1),10003(test-3),10002(test-2)
```

**Comment by treydock** [comment]
Can you provide what some of the corresponding LDAP data looks like?  We use OpenLDAP 2.4 and for example I have dozens os groups and they all show up here:

```
# kubectl get configmap user-gids-map -o yaml -n k8-ldap-configmap | grep tdockendorf
  user-tdockendorf: '["1021","2399","3241","3285","3309","4391","4496","4547","4548","5087","5301","5353","5356","5358","5509","5527","5607","6393","6557","6558","6865","6951","6952","6957","7175","7396","7455","65533"]'
  user-tdockendorftest: '["3378","6606","7477"]'
```

We also use this with Kyverno and OnDemand and so OnDemand will do UID/GID lookup in Ruby and add supplementalGroups that is all GIDs for a user and Kyverno validates that against the configmap so we know it's accurate.
This is my Helm values for OpenLDAP:

```
---
ldapUrl: ldaps://OMIT:636
ldapGroupBaseDN: dc=osc,dc=edu
ldapUserBaseDN: ou=People,dc=osc,dc=edu
ldapGroupFilter: "(&(objectclass=oscGroup)(status=ACTIVE)(cn=P*))"
ldapUserFilter: "(&(objectclass=oscUser)(!(nsAccountLock=TRUE))(loginDisabled=FALSE))"
mappers:
- user-uid
- user-gid
- user-groups
- user-gids
mappersGroupFilter:
- user-gids=(objectclass=posixGroup)
userPrefix: user-
interval: 5m
nodeSelector:
  node-role.kubernetes.io/infrastructure: ''
```

**Comment by 18707602767** [comment]
We use FreeIPA 4.6.8 for the identification and authorization. 
```
# ipa --version
  VERSION: 4.6.8, API_VERSION: 2.237
```
The following is the relevant data information of user zhangrq in freeipa, and the display of its data in configmap user-groups-map
```
# ipa user-show zhangrq --all
  dn: uid=zhangrq,cn=users,cn=accounts,dc=pl,dc=hpc
  User login: zhangrq
  First name: rq
  Last name: zhang
  Full name: rq zhang
  Display name: rq zhang
  Initials: rz
  Home directory: /home/zhangrq
  GECOS: rqzhang@pl.hpc
  Login shell: /bin/sh
  Principal name: zhangrq@PL.HPC
  Principal alias: zhangrq@PL.HPC
  Kerberos principal expiration: 20230826000000Z
  User password expiration: 20230130082501Z
  Email address: zhangrq@pl.hpc
  UID: 1021600018
  GID: 10022
  User authentication types: password
  Account disabled: False
  Preserved user: False
  Password: True
  Member of groups: test-3, test-2, ipausers, test-4, test-1
  Kerberos keys available: True
  ipauniqueid: 97e5aa58-a077-11ed-9200-0242ac120002
  krbextradata: AALdftdjcm9vdC9hZG1pbkBQTC5IUEMA
  krblastpwdchange: 20230130082501Z
  mepmanagedentry: cn=zhangrq,cn=groups,cn=accounts,dc=pl,dc=hpc
  objectclass: top, person, organizationalperson, inetorgperson, inetuser, posixaccount, krbprincipalaux,
               krbticketpolicyaux, ipaobject, ipasshuser, ipaSshGroupOfPubKeys, mepOriginEntry,
               ipauserauthtypeclass

## the group data of user zhangrq in configmap user-groups-map.
# kubectl get cm user-groups-map -o yaml | grep zhangrq
  zhangrq: '["test-4"]'
```

**Comment by treydock** [comment]
When the `ldapMemberScheme=member` the group membership for users is based on the group's `member` attribute.  What does the group member list look like for `test-1` group, for example?  Seeing `ldapsearch` output is more useful since that will more closely map to what this project uses since everything is just LDAP queries.

The only other thing I can think would be an issue is if FreeIPA uses paging so it only returns a small number of results per query rather than all results.  I use OpenLDAP where there is no paging enabled.  You can verify with an `ldapsearch` on `(objectClass=posixGroup)` and see if it results all groups of only some of the groups in FreeIPA.  If only a partial result is returned try adding `--ldap-paged-search` flag which needs to be added with `extraArgs` if using Helm.

**Comment by 18707602767** [comment]
Querying the test-1 group member list information through the ldapsearch command is in line with expectations.	
```
# ldapsearch  -vvv -b  "cn=test-1,cn=groups,cn=accounts,dc=pl,dc=hpc"
  ldap_initialize( <DEFAULT> )
  SASL/GSSAPI authentication started
  SASL username: admin@PL.HPC
  SASL SSF: 256
  SASL data security layer installed.
  filter: (objectclass=*)
  requesting: All userApplication attributes
  # extended LDIF
  #
  # LDAPv3
  # base <cn=test-1,cn=groups,cn=accounts,dc=pl,dc=hpc> with scope subtree
  # filter: (objectclass=*)
  # requesting: ALL
  #
  
  # test-1, groups, accounts, pl.hpc
  dn: cn=test-1,cn=groups,cn=accounts,dc=pl,dc=hpc
  objectClass: top
  objectClass: groupofnames
  objectClass: nestedgroup
  objectClass: ipausergroup
  objectClass: ipaobject
  objectClass: posixgroup
  gidNumber: 10000
  cn: test-1
  ipaUniqueID: c7797ac2-8d9b-11ed-97d0-0242ac120002
  member: uid=zhangran1,cn=users,cn=accounts,dc=pl,dc=hpc
  member: uid=zhangrq,cn=users,cn=accounts,dc=pl,dc=hpc
  
  # search result
  search: 4
  result: 0 Success
  
  # numResponses: 2
  # numEntries: 1
```
Using the ldapsearch command can query all groups in freeipa, and no missing groups are found.In order to prevent paging in the future, the --ldap-paged-search flag is also added. After re-updating, the data display of configmap user-groups-map has not changed.
```
# ldapsearch  -LLL -b  "cn=groups,cn=accounts,dc=pl,dc=hpc" filter "(objectClass=posixGroup)" 
  SASL/GSSAPI authentication started
  SASL username: admin@PL.HPC
  SASL SSF: 256
  SASL data security layer installed.
  dn: cn=groups,cn=accounts,dc=pl,dc=hpc
  
  dn: cn=admins,cn=groups,cn=accounts,dc=pl,dc=hpc
  
  dn: cn=ipausers,cn=groups,cn=accounts,dc=pl,dc=hpc
  
  dn: cn=trust admins,cn=groups,cn=accounts,dc=pl,dc=hpc
  
  dn: cn=test-1,cn=groups,cn=accounts,dc=pl,dc=hpc
  
  dn: cn=test-2,cn=groups,cn=accounts,dc=pl,dc=hpc
  
  dn: cn=test-3,cn=groups,cn=accounts,dc=pl,dc=hpc
  
  dn: cn=test-4,cn=groups,cn=accounts,dc=pl,dc=hpc
  
  dn: cn=zhangran1,cn=groups,cn=accounts,dc=pl,dc=hpc
  
  dn: cn=zhangrq,cn=groups,cn=accounts,dc=pl,dc=hpc



# kubectl exec -it -n k8-ldap-configmap k8-ldap-configmap-6c8c4cf94b-n5d52 -- /bin/sh
  ~ $ ps -ef 
  PID   USER     TIME  COMMAND
  1 nobody    0:00 /k8-ldap-configmap --ldap-url=ldap://10.49.18.102:389 --no-ldap-tls --ldap-tls-verify --ldap-group-base-dn=cn=groups,cn=accounts,dc=pl,dc=hpc --ldap-user-base-dn=cn=users,cn=accounts,dc=pl,dc=hpc --ldap-bind-dn=uid=rancher,cn=sysaccounts,cn=etc,dc=pl,dc=hpc --ldap-group-filter=(objectClass=posixGroup) --ldap-user-filter=(objectClass=posixAccount) --ldap-member-scheme=member --ldap-user-attr-map=name=uid,uid=uidNumber,gid=gidNumber --ldap-group-attr-map=name=cn,gid=gidNumber --mappers=user-uid,user-gid,user-gids,user-groups --namespace=default --interval=5s --listen-address=:8080 --ldap-paged-search --ldap-paged-search-size=2000 --log-level=debug



#  kubectl get cm  user-groups-map -o yaml | grep zhangrq
  zhangrq: '["test-4"]'

```

**Comment by treydock** [comment]
Another thing to try is setting `ldapMemberScheme=memberof` if FreeIPA supports the memberOf plugin to set memberships on the user.  It looks like it might based on one of the outputs where it lists the groups a user is a member of.  Looking at what I run at OSC we are using the `memberof` mapping so could be a bug in the `member` mapping.

Running locally at OSC using `member` I can reproduce the issue.

Using `memberof`:

```
# kubectl get cm user-groups-map -o yaml -n k8-ldap-configmap | grep tdockendorf
  user-tdockendorf: '["PZS0708","PAS1936","PDE0001","PZS1118","PDE0006","PDE0009"]'
```

Writing configmap to different namespace using `member`:

```
# kubectl get cm user-groups-map -o yaml | grep tdockendorf
  tdockendorf: '["PDE0009"]'
  user-tdockendorf: '["PZS0708","PDE0009"]'
```

So looks like there are 2 issues with the `member` mapping, it's causing duplicate entries when using the username prefix and also not populating all the groups.

**Comment by treydock** [comment]
This should be fixed with version v0.7.1
